docs: remove activatedRouteData backward accesses in aio guides (#45140)
angular/angular: aio/content/guide/route-animations.md

After configuring the routes, add a `<router-outlet>` inside the root `AppComponent` template. The `<router-outlet>` directive tells the Angular router where to render the views when matched with a route.

> The `<router-outlet>` directive holds the custom data set for the currently active route which can be accessed via the route's `activatedRouteData` property, we can use such data to animate our routing transitions.

Reviewer: ```suggestion
The `ChildrenOutletContexts` holds information about outlets and activated routes. We can use the `data` property of each `Route` to animate our routing transitions.
```

Here, the `prepareRoute()` method takes the value of the outlet) and returns a string value representing the state of the animation based on the custom data of the current active route. Use this data to control which transition to execute for each route.

<div class="alert is-helpful">

  The `router-outlet` directive holds a `activatedRouteData` property which can be accessed by assigning a template variable to the directive and using that to access the directive's data.

  Like so for example:
  ```html
    <div [@routeAnimations]="prepareRoute(outlet)">
      <router-outlet #outlet="outlet"></router-outlet>
    </div>
  ```
  ```typescript
    prepareRoute(outlet: RouterOutlet) {
      return outlet?.activatedRouteData?.['animation'];
    }
  ```

  That is not however the correct way to access the outlet's data since, it is accessing the outlet before it is defined (as in, the `outlet` usage is before the `outlet` definition in the template), this is called __backward reference__ and can cause various issues since it is not supported by Angular.

> </div>

Reviewer: We don't generally document what _not_ to do but instead just document the correct way of doing things. There are a million different ways to do something incorrectly and we shouldn't attempt to document them.

Author: mh... yeah you do make a good point, I just feel that this is so common and so much used that it may be better to upfront let the user know that this is not the best way to do it :thinking: 

For example at the end of the animations guide we provide a link to a angular-connect talk ([here](https://angular.io/guide/animations#more-on-angular-animations)), even in the talk itself they show this way of accessing the outlet data: [video at 16:56](https://youtu.be/rnTK9meY5us?t=1016)

Lots of other places use/teach this way, like this article I read a while back: [indepth angular animations - route animations](https://indepth.dev/posts/1285/in-depth-guide-into-animations-in-angular#route-animations), this [video tutorial](https://youtu.be/7JA90VI9fAI?t=107) this [medium post](https://medium.com/ngconf/animating-angular-route-transitions-ef02b871cc30) and the list can go on and on I think :confused: 

So yeah if you really dislike it I can remove it, or I could also wrap the code in [`avoid` code snippets](https://angular.io/guide/docs-style-guide#displaying-bad-code)

Please let me know what you prefer :slightly_smiling_face:

Author: I see :slightly_smiling_face: , ok let me remove that then :slightly_smiling_face: 

About breaking changes, sorry I didn't think this was something that was ok in the past and not anymore (ah... I guess it's related on the way Ivy parses the templates? :thinking:)

I really like the angular connect video, so I wouldn't really mind keeping it there (all that is says is still relevant/current in the animations api), but yeah I agree that it's sort of presents the same info but just in video format :+1: (but if you want I can open a PR to remove it :slightly_smiling_face:)

I did search "backward reference" in the repoand nothing in the docs popps up so it's either called something different or it's not mentioned at all:
![indipendent-on-open](https://user-images.githubusercontent.com/61631103/154770569-59a8a618-5062-4a52-b38f-060f757c402b.gif)

If you want I'd like to add something in the docs around this :slightly_smiling_face:

Reviewer: In the [v9 compatibility](https://v9.angular.io/guide/ivy-compatibility) docs, it was noted as "Forward references to directive inputs accessed through local refs are no longer supported by default." In other internal docs, we had it described as "backwards references". I know this mentions inputs specifically, but I don't think it's really limited to that (but rather, things that change during the execution of the component, which is more of a mouthful and harder to explain and document). 

I think this generally falls into the "unidirectional data flow" concept. I'm not super familiar with the entirety of our docs so I don't know if this is even mentioned specifically.

Author: I see, thanks, I think I will have a look around to see if I can find it somewhere, and if not I'll see if I can find a good place to add it in :slightly_smiling_face::+1:

Author: ah yeah, sorry I forgot about that... yeah hopefully people can see that the video is a 2017 one so may be cautious that it may be slightly out of date... :thinking: 

not sure, if you want I think we can totally remove it, I really like it as it is indeed well presented and engaging, but yeah it is indeed not ideal that it mentions that way of accessing the route data :sweat:

I am sort on the fence I think I might prefer to keep it in, but just slightly (it also inconsistent with other guides as far I can tell, not many guides present videos of conferences)

:shrug:

Reviewer: You like it so let's keep it :)

Author: :smile::+1:
